mansrc/compileSpliceCands.1
'\" t
.\"     Title: compilesplicecands
.\"    Author: [see the "AUTHORS" section]
.\" Generator: Asciidoctor 1.5.5
.\"      Date: 
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "COMPILESPLICECANDS" "1" "" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.LINKSTYLE blue R < >
.SH "NAME"
compilesplicecands \- Find introns parsing a set of exon candidates in GFF format
.SH "SYNOPSIS"
.sp
\fBcompileSpliceCands\fP \-f <splice candidates filename> \-g <genome filename>
.SH "OPTIONS"
.SS "Parameters:"
.sp
\fB\-f\fP <filename>
.RS 4
potential splice sites in gff format
.RE
.sp
\fB\-g\fP <filename>
.RS 4
genome file in fasta format
.RE
.sp
\fB\-c\fP <integer>
.RS 4
defines the number of potential splice site every potential splice site itself is compared with
.RE
.sp
\fB\-t\fP <float>
.RS 4
has to be between 0 and 1 and  defines how much the average coverage may differ
.RE
.sp
\fB\-m\fP <integer>
.RS 4
defines how many bases around the splice site should be checked
.RE
.sp
\fB\-d\fP
.RS 4
enables debugging output
.RE
.sp
\fB\-i\fP <integer>
.RS 4
maximum length an Intron can be, default 500000 (human genome)
.RE
.SH "AUTHORS"
.sp
AUGUSTUS was written by M. Stanke, O. Keller, S. König, L. Gerischer and L. Romoth.
.SH "ADDITIONAL DOCUMENTATION"
.sp
An exhaustive documentation can be found in the file /usr/share/doc/augustus/README.md.gz.

HISTORY.TXT
     - remove bam2wig
     - compile with CGP and ZIPINPUT support by default
     - compile with SQLite and MySQL support for CGP (switch of by adding SQLite = false and MySQL = false to common.mk)
List of changes from version 3.3.2 to 3.3.3 (until Sep 13th, 2019)
     - docker container
     - new script tfix_in_frame_stop_codon_genes.py that replaces genes where spliced in-frame stop codons were predicted
     - new scripts: compare_masking.pl, merge_masking.pl
     - new program bamToWig.py as alternarive to bam2wig C++ binary
     - fix warnings on new GCC compiler (8.3)
     - introduced unit tests of C++ code (make test)
     - bugfixes (lex.cc, autoAug.pl)
     - DIAMOND as alternative to BLAST in aa2nonred.pl
List of changes from version 3.3.1 to 3.3.2 (until Oct 5th, 2018)
     - bugfixes in comparative augustus, utrrnaseq
     - new species Chiloscyllium punctatum (bamboo shark), Scyliorhinus torazame (cat shark), Rhincodon typus (whale shark)
     - updated comparative augustus (CGP) tutorial 
List of changes from version 3.3 to 3.3.1 (until May 8th, 2018)
     - new species pisaster (Pisaster ochraceus, ochre starfish)
     - bugfixes of sampling error in intron model and in check of
       frame-compatibility of CDSpart hints without frame
     - new auxiliary program utrrnaseq to find training UTRs
     - new script setStopCodonFreqs.pl
List of changes from version 3.2.3 to 3.3 (until July 11th, 2017)
     - new program ESPOCA to estimate selective pressure on codon alignments
     - gene finding on ancestral genomes is enabled
     - new default parameters for comparative gene prediction (CGP)
     - clade parameters training for CGP
     - compatibility to Ian Fiddes' Comparative Annotation Toolkit (CAT)
     - new scripts eval_dualdecomp.pl,
     - more tolerant tree parsing
     - bugfixes in augustus, joingenes, load2sqlitedb, transMap2hints.pl, splitMfasta.pl, intron2exex.pl,
       aln2wig
     - new functionality in homGeneMapping, joingenes
     - new species ciona, mnemiopsis_leidyi, nematostella_vectensis, strongylocentrotus_purpuratus
List of changes from version 3.2.2 to 3.2.3 (until December 6th, 2016)
     - bugfixes in joingenes, load2sqlitedb, transMap2hints.pl
     - added script gtf2bed.pl
     - CGP parameters for clade of 8 primates (cgp/log_reg_parameters_primates.cfg)
     - added option --clean to load2sqlitedb
     - now filter out truncated coding genes with CDS length <= 3
     - Sascha Steinbiss: remove dependency on PATH_MAX
List of changes from version 3.2.1 to 3.2.2 (until March 30th, 2016)
     - added extrinsic.M.RM.PB.cfg for PacBio CCS hints with GMAP
       (see http://bioinf.uni-greifswald.de/bioinf/wiki/pmwiki.php?n=Augustus.PacBioGMAP )
     - standardization of Makefile
     - improvements in auxiliary program joingenes
     - added missing program load2sqlitedb
     - comparative gene predictions (CGP):
         - hints with source=M (e.g. existing annotations) are fixed
	 - new standard parameters for exon_gain, exon_loss and others
     - bugfixes, in particular a double free or corruption bug in PPX that occured with BUSCO
     - new species Volvox carteri (volvox)
List of changes from version 3.2 to 3.2.1 (until October 23rd, 2015)
     - comparative gene prediction (CGP): make the parameters from the GCB proceedings paper match again
     - new CGP default parameters (tested on 5 clades)
     - CGP version penalizes very long introns equivalently to a geometric distribution
     - code now under artistic license only
List of changes from version 3.1 to 3.2 (until September 25th, 2015)
     - comparative gene prediction (CGP option) has sensible default parameters of the logistic regression model
     - new tool hal2maf_split.pl for splitting an alignment for parallel comparative prediction
     - bugfixes to augustus (compiler warnings on MaxOS)
     - improvements to prediction of noncoding genes
     - homGeneMapping can determine homologous sets of transcripts
     - bugfixes in joingenes and bam2hints (rare segmentation fault)
     - made load2sqlitedb more robust, new option to skip index building when loading many species
     - new tutorial that includes iterative training
     - new option to shift coordinates in hint input when sequence window is cut out
     - adapted cegma2gff.pl to cegma version 2.5
List of changes from version 3.0.3 to 3.1 (until April 29th, 2015)
     - prediction of noncoding genes (ncRNA) with option --nc=on, requires input of hints, e.g. from RNA-Seq,
       predicts coding genes (with UTRs) and noncoding genes simultaneously
     - speed improvement with --UTR=1 (by 83% on human without sampling, by 18% on fly with sampling)
     - softmasking support: --softmasking=1 is a little more accurate and much faster on human
     - small overall accuracy improvement
     - prevent the prediction of spliced in-frame stop codons (for SHORT or HINTED introns only)
     - new species: rice, Danio rerio (zebrafish), s_aureus (Staphylococcus aureus),
       camponotus_floridanus (ant) contributed by Shishir K Gupta,
       sulfolobus_solfataricus (archaeon), Parasteatoda tepidariorum (house spider), Apis dorsata (from Francisco Camara Ferreira)
     - update to parameters of Thermoanaerobacter tencongensis, E.coli, Yarrowia lipolytica, schistosoma2 (Schistosoma mansoni)
     - remove dependency on libboost-filesystems, boost is now completely optional in single-species mode
     - new auxiliary program 'homGeneMapping' for identification of homologous genes in a set of gene predictions
     - new auxiliary program 'joingenes' for combining gene sets
     - bugfixes in load2db, length distribution skew of bacteria, comparative mode (findGeneRanges,
       Alignment::mergeable, memory leak), some ass hints were previously not used although possible
     - decreased minintronlen default of bam2hints to 32 as the old threshold tended to be too large
     - computing of omega = dN/dS more efficiently

auxprogs/compileSpliceCands/compileSpliceCands.c
/*******************************************************************************/
/* compileSpliceCands: 			                                       */
/*                                                                             */
/* Autor: Ralph Krimmel                                                        */
/* email: thirsty@milk-and-cookies.net                                         */
/*******************************************************************************/




#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include "list.h"


typedef int bool;
#define TRUE   (1)
#define FALSE  (0)

#define FILEBUFFER 1048576
#define LINEBUFFER 256

/*variable for debug output*/
static int debug=0;

 /* chomp a string*/
void chomp(const char *s)
{
	char *p;
	p = strchr(s, '\n'); if (p != NULL) *p = '\0';
}

/*prints out an error message when allocating memory fails and exits*/
void printOomError()
{
	printf("Could not allocate enough memory, exiting..\n");
	fflush(stdout);
	exit(1);
}

/*returns the tokens of a string split by the separator in an array of char **/
int strtoken(char ***target, char *in, char sep) {
	int ret = 0;
	int i = 0;
	int si = 0;
	char **out = *target;
	while (in[i] != '\0') 
	{
		if ((in[i] != sep)) 
		{
   			out[ret][si++] = in[i];
 		} 
		else 
		{
   			out[ret][si] = '\0';
   			si = 0;
   			ret++; 
		}
 		i++;
	}
	out[ret][si] = '\0';
	if (i > 0) 
	{
		ret++;
	}
	

 	return ret;
}

/*checks if the input is still sorted. not used at the moment*/
int stillSorted(char ***run,char *chkstr,int strcount)
{
	int i;
	char **target = *run;
	for (i=0; i < strcount; i++) 
	{
		if (strcmp(target[i], chkstr) == 0)
		{
		        fprintf(stderr, "Found %s!. This should never happen on sorted input data.\nPlease sort the input.\nYou can do this by using the GNU sort tool.\n", *target);
			return FALSE;
		}
        }
	return TRUE;
}

/*checks two strands which can be +,- and . if they are compatible.
 * compatible strands are '++', '--', '+.', '.+', '-.', '.-' and '..'  */
int isCompatibleStrand(char strand1, char strand2)
{
	if((strand1==strand2) || (strand1=='.') || (strand2 == '.'))
	{
		return TRUE;
	}
	return FALSE;
}

/*parses a double out of a string formatted like this: grp0-0-22.1373 
 * The return value would be 22.1373 then */
double getAverageCoverage(char *s)
{
	int i=0;
	char *pch;

	pch = strtok (s,"-");
	while (pch != NULL)
	{	
		i++;
		pch = strtok (NULL, "-");
		if(i==2)
		{	
			return atof(pch);
		}
		
 	}
	return 0.0;
}

/*returns the reverse complement of a char*/
char rc(char in)
{
	switch(in)
	{
		case 'a':
		  return 't';
		case 't':
		  return 'a';
		case 'c':
		  return 'g';
		case 'g':
		  return 'c';
		default:
		  return in;
	}
}

/*searches on a genome in two ranges for splice sites and prints the resulting introns in gff format
 * if there are unknown strands, it searches in both directions. This function should just be called with compatibleStrand arguments
 * (see isCompatibleStrand function) */
void findPossibleSpliceSites(char *chromosomeName,char *genomePart, long int startPosLeft, long int endPosLeft, long int startPosRight, long int endPosRight, char strandLeft, char strandRight)
{
	long int i,j;
	char resultingStrand;
	if(debug) printf("startPosLeft: %ld: | endPosLeft: %ld | startPosRight: %ld | endPosRight: %ld |strandleft: %c | strandRight: %c\n",startPosLeft,endPosLeft,startPosRight,endPosRight,strandLeft,strandRight);
		
	resultingStrand = (strandLeft==strandRight) ? strandLeft : '.';
	if((strandLeft=='+') || strandRight=='+' || (( strandLeft=='.') && (strandRight=='.' )))
	{
		for(i=startPosLeft;i<endPosLeft-1;i++)
			{
			if((genomePart[i]=='c') && (genomePart[i+1] =='t') )
			{
				for (j=startPosRight;j<endPosRight-1;j++)
				{
					if((genomePart[j] == 'a') && (genomePart[j+1] == 'g'))
					{
						printf("%s\t%s\t%s\t%ld\t%ld\t%i\t%c\t%i\t%s\n",chromosomeName,"source","intron",i,j,99,resultingStrand,1,"");
					}
				}	
			}
		}
	}
	if((strandLeft=='-') || (strandRight=='-') || ((strandLeft=='.') && (strandRight=='.' )))
	{
		for(i=endPosLeft;i>startPosLeft+1;i--)
		{
			if(((genomePart[i])=='g') && (genomePart[i-1] == 'a'))
			{
				for (j=endPosRight;j>startPosRight+1;j--)
				{
					if((genomePart[j]=='t') && (genomePart[j-1]=='c'))
					{
						printf("%s\t%s\t%s\t%ld\t%ld\t%i\t%c\t%i\t%s\n",chromosomeName,"source","intron",i,j,99,resultingStrand,1,"");
					}
				}
			}
		}
	}
}


/*searches a needed genome given by chromosomename and returns is as string*/
char *getChromosome(char *fn,char* chromosomename)
{
	/*the file handle that hold the genome file*/
	FILE *genome_handle=NULL;
	/*the buffer that holds a line while reading the genome*/
        char line[LINEBUFFER];
	/*the return string*/
        char *s=NULL;
	/*the length of the actual read string*/
        int line_length=0;
	/*when the fasta description of the next chromosome is found done is set to true*/
        int done=FALSE;
	/*true when the wanted chromosome is found. from this point on, the file is read into ram*/
	int startreading=FALSE;
	/*the string we are searching for in the genome (means: ">chromosomename")*/
	char *checkstring=NULL;
	/* the actual position we want to write the next char of the genome to*/ 
	long int position=0;
	int i=0;
	if(debug) 
	{
		fflush(stdout);
		printf("Trying to get %s from file\n",chromosomename);	
	}
	/*assemble the checkstring*/
	checkstring=calloc(strlen(chromosomename)+2,1);
	if(checkstring==NULL) printOomError();

	checkstring[0]='>';
	strncat(checkstring,chromosomename,strlen(chromosomename));
	
	/*open the genome file*/
	genome_handle = fopen ( fn, "r" );
        if (genome_handle == NULL)
        {
                printf("Could not open genome file\n");
        }
        
	/*get enough memory to hold the chromosome*/
	s = malloc(3500000000);
	if(s == NULL) printOomError();
	
	/*read in the chromosome*/
	while ((fgets ( line, sizeof line, genome_handle ) != NULL) && !(done))
	{
		chomp(line);
		if(strcmp(line,checkstring)==0)	
		{
			startreading=TRUE;
			if(debug) 
			{
				printf("Found %s, reading it into ram\n",chromosomename);
				fflush(stdout);
			}
			continue;
		}	
		if(startreading)
		{

			line_length=strlen(line);
			if(line[0]=='>')
			{
				done=TRUE;
			}
			else
			{
				for(i=0;i<line_length;i++)
				{
					s[i+position]=tolower(line[i]);
				}
				position+=line_length;
			}
		}
	}
	free(checkstring);
	fclose(genome_handle);
	if(debug) printf("Successfully loaded chromosome %s into memory\n",chromosomename);
	fflush(stdout);
	return s;
}

/*parses a line and returns a filled Data struct */
Data parseLine(char *ln)
{
	Data d = malloc(sizeof(struct LineData));

	char **tokens=NULL;
	int i;	
	tokens = malloc(9*sizeof(char*));
	if (tokens ==NULL) 
	{
		printf("Couln't allocate memory, exiting\n");
		exit(1);
	}
			for (i = 0; i < 9; i++) 
	{
		tokens[i] = malloc(LINEBUFFER);
		if(tokens[i] == NULL)
		{	
			printf("Couln't allocate memory, exiting\n");
			exit(1);
		}
	}

	strtoken(&tokens,ln,'\t');
	strncpy(d->name,tokens[0],LINEBUFFER);
	d->startPos=atoi(tokens[3]);
	d->endPos=atoi(tokens[4]);
	d->strand=tokens[6][0];
	d->averageCoverage=getAverageCoverage(tokens[8]);
	
	for(i=0;i<9;i++) 
	{
		free(tokens[i]);
		tokens[i]=NULL;
	}
	free(tokens); 
	tokens=NULL;
return d;
}

/*parses the linked list of Data structs */
int parseList(List *L, char *actualChromosome,long int chromosomeLength,char* oldname,int maxSpliceSiteDiff,float threshold,int maxIntronLength)
{
	List linkedList = *L;
	Position P;

	Data left;
	Data right;
	int isCandidate;
	signed int leftBorder;
	long int rightBorder;
	

	P=Header(linkedList);
	if(debug) printf("parsing list\n"),fflush(stdout);
	while  (!IsLast(P,linkedList))
	{
		P=Advance(P);
		if(P == First(linkedList))
		{
			left=Retrieve(P);	
		}
		else
		{
			right=Retrieve(P);
			isCandidate = ((right->averageCoverage>=left->averageCoverage*threshold) || 
				       (left->averageCoverage*((1/threshold))<right->averageCoverage));
			if(isCandidate && (maxIntronLength > (right->endPos-left->startPos)))
			{	
				if(isCompatibleStrand(left->strand,right->strand))
				{
					leftBorder=left->startPos-maxSpliceSiteDiff;
					
					rightBorder=right->endPos+maxSpliceSiteDiff;
					if ((leftBorder>0) && (rightBorder<chromosomeLength))
					{
						findPossibleSpliceSites(left->name,actualChromosome,left->startPos-maxSpliceSiteDiff,left->endPos+maxSpliceSiteDiff,right->startPos-maxSpliceSiteDiff,right->startPos+maxSpliceSiteDiff,left->strand,right->strand); 
					}
					else
					{
						if((left->startPos>0) && (right->endPos<chromosomeLength))
						{
							findPossibleSpliceSites(left->name,actualChromosome,left->startPos,left->endPos,right->startPos,right->endPos,left->strand,right->strand);
						}
					}

				}
			}
		}
	}
	return 0;
}


void printList(const List L) 
{
	Position P = Header(L);
	if ( IsEmpty(L) ) 
	{
		printf("[Empty list]\n");
	} 
	else 
	{
		do 
		{
      			P = Advance(P);
			printf("name:%s| avgc:%f | start: %ld | end: %ld\n", Retrieve(P)->name,Retrieve(P)->averageCoverage,Retrieve(P)->startPos,Retrieve(P)->endPos);
    		} while (!IsLast(P, L));
	}
	fflush(stdout);
}



int main (int argc, char **argv)
{
	char *gff_filename=NULL;
	char *genome_filename=NULL;
	/*char *trackname=NULL; */
	
	FILE *gff_handle=NULL; 

	char usage[]="\ncompileSpliceCands: Find introns parsing a set of exon candidates in gff format. USAGE: compileSpliceCands -f <splice candidates filename> -g <genome filename>\n\nOption:\t\tArgument:\tDescription:\n\t-f\t<filename>\tThe potential splice sites in gff format\n\t-g\t<filename>\tThe genome file in fasta format)\n\t-c\t<integer>\tDefines the number of potential splice site every potential splice site itself is compared with\n";
	char usage1[]="\t-t\t<float>\t\thas to be between 0 and 1 and defines how much the average coverage may differ\n\t-m\t<integer>\tdefines how many bases around the splice site should be checked.\n\t-d\t<no argument>\tenables debugging output\n\t-i\t<integer>\tThe maximum length an Intron can be. Default 500000(human genome).\n";
	
	char buffer[FILEBUFFER];
	
	char line[LINEBUFFER];

	/* the name of the last chromosome */	
        char oldname[LINEBUFFER];
	
	int maxSpliceSiteDiff=10;
	
	float threshold=0.6;

	int checkWidth=3;

	int i=0;
	
	int c=0;

	char *actualChromosome=NULL;

	int listCount=0;

	int sameChromosome=TRUE;
	
	long int chromosomeLength;

	int maxIntronLength=500000;

	List L=NULL;
	
	Position P=NULL; 

	Data tmpData;

	opterr = 0;
	
	if ( argc < 5  ) 
	{
		printf ("%s%s",usage,usage1);
		exit(2);
	}

	/*Parse command line arguments*/
	while ((c = getopt (argc, argv, "f:g:c:t:m:i:d")) != -1)
	        switch (c)
		{
		case 'f':
			gff_filename = optarg;
			break;
		case 'g':
			genome_filename = optarg;
			break;
		case 'c':
			checkWidth = atoi(optarg);
			break;
		case 'm':
			maxSpliceSiteDiff = atoi(optarg);
			break;
		case 'd':
			debug=1;
			break;
		case 'i':
			maxIntronLength = atoi(optarg);
			break;
		case 't':
			threshold = atof(optarg);
			break;
		case '?':
			if (optopt == 'c')
		  		fprintf (stderr, "Option -%c requires an argument.\n", optopt);
			else if (isprint (optopt)) 
			{
		  		fprintf (stderr, "Unknown option `-%c'.\n", optopt);
				fprintf (stderr, "%s%s",usage,usage1);
		  	}
		  	else
		  	{
				fprintf (stderr,"Unknown option character `\\x%x'.\n",optopt);
				fprintf (stderr, "%s%s",usage,usage1);
				return 1;
		  	}
		default:
			fprintf(stderr, "Don't know what to do, exiting\n");
			abort ();
           	}
	
	
	/* invalid command line arguments */
	if(checkWidth == 0) 
	{
		checkWidth=4;
	}
	if(maxSpliceSiteDiff == 0)
	{
		maxSpliceSiteDiff=10;
	}
	if(threshold == 0.0)
	{
		threshold = 0.6;
	}
	if(maxIntronLength == 0)
	{

	}	maxIntronLength = 500000;

	/*cause we start counting from zero ;) */
	checkWidth++;
	gff_handle = fopen ( gff_filename, "r" );
	if ( gff_handle != NULL )
	{
		/*create an empty linked list and set the position variable to the header*/	
		 L = MakeEmpty(NULL);
		 P = Header(L);
		
		/*we want buffered input */
		setbuf(gff_handle, buffer);

		i=0;
		/*initial fill of the linked list*/
		while((fgets ( line, sizeof line, gff_handle ) != NULL)) 
		{
			chomp(line);
			tmpData=parseLine(line);
			if(i==0)
			{
				actualChromosome=getChromosome(genome_filename,tmpData->name);
				chromosomeLength=strlen(actualChromosome);
				strncpy(oldname,tmpData->name,LINEBUFFER);
				i++;
			}
			sameChromosome=((strcmp(tmpData->name,oldname)==0));
			if(debug) 
			{
				printf("tmpData->name=%s tempData->averageCovergage:%f| oldname:%s\n| listcount: %i | checkWidth: %i  | sameChromosome: %i\n",tmpData->name,tmpData->averageCoverage,oldname,listCount,checkWidth,sameChromosome);
				printList(L);
				printf("\n");
				fflush(stdout);
			}
			if((listCount<checkWidth && sameChromosome))
			{
				Insert(tmpData,L,P);
				P=Advance(P);
				listCount++;
			}
			else
			{
				if((listCount==checkWidth && sameChromosome))
				{
					parseList(&L,actualChromosome,chromosomeLength,oldname,maxSpliceSiteDiff,threshold,maxIntronLength); 
					Delete(Retrieve(First(L)),L);
					Insert(tmpData,L,P);
					P=Advance(P);
				}

				else
				{
					parseList(&L,actualChromosome,chromosomeLength,oldname,maxSpliceSiteDiff,threshold,maxIntronLength); 
					free(actualChromosome);
					actualChromosome=getChromosome(genome_filename,tmpData->name);
					chromosomeLength=strlen(actualChromosome);
					strcpy(oldname,tmpData->name);
					L=MakeEmpty(L);
					P=Header(L);
					Insert(tmpData,L,P);
					P=Advance(P);
					listCount=1;
				}
			}
		}

		if(debug) printList(L);	
		parseList(&L,actualChromosome,chromosomeLength,oldname,maxSpliceSiteDiff,threshold,maxIntronLength); 
		printf("\n");
	 	fclose ( gff_handle );
		DeleteList(L); 
	}
	else
	{
		perror ( gff_filename ); /* why didn't the file open? */
		printf("%s%s",usage,usage1);
	}
	free(actualChromosome);
	return 0;
}
